Generating DEB and RPM files with the source from the rocm-examples.

Mustaballer opened the pull request:
Adds a script to automate the packaging of ROCm SDK examples into DEB and RPM packages. The script performs the following tasks:
- Sets up necessary package metadata and directories.
- Copies source files and directories to the build directory.
- Creates DEB and RPM packages with the specified metadata and source files.
- Cleans up previous build artifacts to ensure a fresh build environment.

Mustaballer committed:
copy changes from #131

Mustaballer committed:
exclude binary, object and executable files

Mustaballer committed:
test "bin/*" exclusion

Mustaballer committed:
add build_packages.sh script to generate deb and rpm files from source

Mustaballer committed:
fix issues with rpm file generation

Mustaballer commented:
The rpm file generation also seems to be working, you can view file contents of the rpm file generated from `build_packages.sh` here 
[rpm-file-contents.txt](https://github.com/user-attachments/files/16275099/rpm-file-contents.txt)

Mustaballer committed:
revert changes to CMakeLists.txt

Mustaballer commented:
@dgaliffiAMD PR is ready for review! :)

Mustaballer committed:
add directive ective that prevents the RPM build process from terminating due to missing build ID

dgaliffiAMD commented on build_packages.sh:
```diff
@@ -0,0 +1,144 @@
+PACKAGE_DESCRIPTION_SUMMARY="A collection of examples for the ROCm software stack"
+PACKAGE_INSTALL_PREFIX="/opt/rocm/examples"
+
+BUILD_DIR="$(pwd)/build"
```
```suggestion
GIT_TOP_LEVEL=$(git rev-parse --show-toplevel)
BUILD_DIR="$(GIT_TOP_LEVEL)/build"
```
To help the script find the top level.

dgaliffiAMD committed:
Set build_packages.sh to executable

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD commented on build_packages.sh:
It would be good to move this script to the `Scripts` folder, with the others. Maybe `Scripts/packaging`.

dgaliffiAMD commented on build_packages.sh:
```diff
@@ -0,0 +1,144 @@
+    rm -rf $RPM_BUILD_DIR $RPM_SOURCE_DIR $RPM_SPEC_DIR $RPM_RPMS_DIR $RPM_SRPM_DIR
+}
+
+# Copy sources to build directory
+copy_sources $BUILD_DIR/${PACKAGE_NAME}-${PACKAGE_VERSION}
+
+# Create DEB package
+create_deb_package $BUILD_DIR/${PACKAGE_NAME}-${PACKAGE_VERSION}
+
+# Create RPM package
+create_rpm_package $BUILD_DIR/${PACKAGE_NAME}-${PACKAGE_VERSION}
```
```suggestion
pushd $GIT_TOP_LEVEL || exit

# Copy sources to build directory
copy_sources $BUILD_DIR/${PACKAGE_NAME}-${PACKAGE_VERSION}

# Create DEB package
create_deb_package $BUILD_DIR/${PACKAGE_NAME}-${PACKAGE_VERSION}

# Create RPM package
create_rpm_package $BUILD_DIR/${PACKAGE_NAME}-${PACKAGE_VERSION}

popd || exit
```
Explicitly change the directory, so that we know we're working from the top level.

dgaliffiAMD commented on build_packages.sh:
```diff
@@ -0,0 +1,144 @@
+    find $RPM_RPMS_DIR -name "${PACKAGE_NAME}-${PACKAGE_VERSION}-*.rpm" -exec mv {} $RPM_DIR \;
+    rm -rf $RPM_BUILD_DIR $RPM_SOURCE_DIR $RPM_SPEC_DIR $RPM_RPMS_DIR $RPM_SRPM_DIR
+}
+
```
```suggestion
# Clean up previous build artifacts
rm -rf $BUILD_DIR
mkdir -p $DEB_DIR $RPM_BUILD_DIR $RPM_SOURCE_DIR $RPM_SPEC_DIR $RPM_RPMS_DIR $RPM_SRPM_DIR

```
Adding this to the main section, so it's easier to see.

dgaliffiAMD commented on build_packages.sh:
```diff
@@ -0,0 +1,144 @@
+    "LLVM_ASAN"
+)
+
+# Clean up previous build artifacts
+rm -rf $BUILD_DIR
+mkdir -p $DEB_DIR $RPM_BUILD_DIR $RPM_SOURCE_DIR $RPM_SPEC_DIR $RPM_RPMS_DIR $RPM_SRPM_DIR
+
```
```suggestion
```
Moved lower down.

Mustaballer committed:
address PR review comments

Mustaballer committed:
Merge branch 'generate-deb-and-rpm-files-with-source' of github.com:ROCm/rocm-examples into generate-deb-and-rpm-files-with-source

dgaliffiAMD committed:
Renamed 'packaging' to 'Packaging'

To match casing of the other subfolders

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

Beanavil commented on Scripts/Packaging/build_packages.sh:
```diff
@@ -0,0 +1,149 @@
+#!/bin/bash
+# MIT License
+#
+# Copyright (c) 2022-2024 Advanced Micro Devices, Inc. All rights reserved.
```
Nit:
```suggestion
# Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```

Beanavil commented on Scripts/Packaging/build_packages.sh:
```diff
@@ -0,0 +1,149 @@
+set -e
+
+PACKAGE_NAME="ROCm-SDK-Examples"
+PACKAGE_VERSION="6.2.0"
```
What do you think about passing this as a parameter? So that we don't need to update this script for every release

dgaliffiAMD replied:
Yes, please. This will be important for internal builds and packaging too. 
PACKAGE_NAME, PACKAGE_VERSION, INSTALL_PREFIX, and BUILD_DIR

Beanavil replied:
I see these have been addressed, closing conversation

Beanavil commented on Scripts/Packaging/build_packages.sh:
```diff
@@ -0,0 +1,149 @@
+    echo "Architecture: amd64" >> $control_file
+    echo "Maintainer: $PACKAGE_CONTACT" >> $control_file
+    echo "Description: $PACKAGE_DESCRIPTION_SUMMARY" >> $control_file
+    echo "Homepage: https://github.com/ROCm/ROCm-examples" >> $control_file
```
Also nit: the RPM uses this value as well, I would define a variable `PACKAGE_HOMEPAGE_URL` or something like that with this link, just to make sure that we set up always the same value for both packages.

Beanavil commented:
Also, this is probably out of the scope of this PR, but AFAIK a GitHub action could be set up to build the packages and add them to the "Releases" section of the repo when, for instance, a release tagged branch `rocm-x.y.z` is pushed:
```yaml
# Release workflow file
on:
  push:
    tags:
      - 'rocm-*'

jobs:
  package:
    steps:
      - name: Build packages
        run: Scripts/Package/build_packages.sh

      - name: Release packages
        uses: softprops/action-gh-release@v1
        with:
          draft: true
          files: /path/to/generated/packages
```

Mustaballer committed:
Refactor build script to accept PACKAGE_VERSION as a parameter and define PACKAGE_HOMEPAGE_URL variable for consistency in DEB and RPM packages.

Mustaballer committed:
disable the generation of debug packages

Mustaballer committed:
add PACKAGE_NAME,
PACKAGE_VERSION,
PACKAGE_INSTALL_PREFIX,
BUILD_DIR as arguments and add default values

dgaliffiAMD committed:
Add DEB_DIR and RPM_DIRas arguments

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD commented:
I'm not seeing the install prefix on the generated DEB file anymore. Everything is being installed directly to `/`. I'm not sure when this started.
Maybe it will help to incorporate the install prefix to you `copy_source`, when you are setting up the "staging area".

Mustaballer committed:
Fix DEB package directory structure

dgaliffiAMD committed:
Removed "PACKAGE_VENDOR" as it is not used.

dgaliffiAMD committed:
Make RPM_*_DIR variables local to create_rpm_package

dgaliffiAMD committed:
Refactor create_deb_package

Using the "cat EOF" to create DEBIAN/control, rather than multiple "echo" statments. Similar to how the RPM spec files is generated.

dgaliffiAMD committed:
Add DEB_PACKAGE_RELEASE and RPM_PACKAGE_RELEASE as arguments.

dgaliffiAMD committed:
Create a STAGING_DIR contant

Use "STAGING_DIR" constant rather than input parameter to helper functions. This matches the style of other constants being used, like BUILD_DIR and SOURCE_DIR

dgaliffiAMD committed:
Print input variables

Useful for build logs.

dgaliffiAMD committed:
Remove the "clean up" at the end of the package generation.

Artifacts can be used for debugging.

dgaliffiAMD committed:
Add a log at each step

dgaliffiAMD committed:
Fix linting errors found with "shellcheck"

dgaliffiAMD committed:
Another linting issue.
